File: csharp/HandlerChain_Message_Class.cs
// See HandlerChain_Class.cs for details of how the classes in this file are
// used.

using System;

namespace DesignPatternExamples_csharp
{
    /// <summary>
    /// Type of message handled by MessageWindow.
    /// </summary>
    public enum MessageType
    {
        /// <summary>
        /// Window is asked to close itself, generally sent by the window itself
        /// in response to a button up in a Close region.  Applies only to the
        /// currently selected window.
        /// </summary>
        Close = 0,

        /// <summary>
        /// Selects a window based on position.
        /// </summary>
        ButtonDown = 1,

        /// <summary>
        /// Take an action on the currently selected window.
        /// </summary>
        ButtonUp,
    }



    //========================================================================
    //========================================================================
    //========================================================================



    /// <summary>
    /// Position of the message in global coordinates (same scope of
    /// coordinates as windows).  Although this can handle negative
    /// coordinates, the example is careful to use only positive coordinates.
    /// </summary>
    public struct MessagePosition
    {
        /// <summary>
        /// X position.
        /// </summary>
        public int X;

        /// <summary>
        /// Y position.
        /// </summary>
        public int Y;

        /// <summary>
        /// Constructor.
        /// </summary>
        /// <param name="x">X position to use.</param>
        /// <param name="y">Y position to use.</param>
        public MessagePosition(int x, int y)
        {
            X = x;
            Y = y;
        }

        /// <summary>
        /// Convert this position to a string.
        /// </summary>
        /// <returns>A string containing a string representation of the position.</returns>
        public override string ToString()
        {
            return String.Format("x={0,2},y={1,2}", X, Y);
        }
    }



    //========================================================================
    //========================================================================
    //========================================================================



    /// <summary>
    /// Represents a message sent to the windows.  A message contains a type
    /// and a position.
    /// </summary>
    public struct Message
    {
        /// <summary>
        /// Value from the MessageType enumeration indicating the type of this message.
        /// </summary>
        public MessageType MessageType;

        /// <summary>
        /// Position of message when the message was sent.  In a real system, this
        /// would generally represent the position of a cursor at the time the message
        /// was generated.
        /// </summary>
        public MessagePosition Position;

        /// <summary>
        /// Constructor.
        /// </summary>
        /// <param name="type">Value from the MessageType enumeration indicating the
        /// type of message being created.</param>
        /// <param name="position">Position of the message, expressed as a MessagePosition
        /// object.</param>
        public Message(MessageType type, MessagePosition position)
        {
            MessageType = type;
            Position = position;
        }

        /// <summary>
        /// Convert this message to a string.
        /// </summary>
        /// <returns>Returns a string representation of this message.</returns>
        public override string ToString()
        {
            return String.Format("{0} at ({1})", MessageType, Position);
        }
    }
}

File: python/DesignPatternExamples_python/handlerchain/handlerchain_message_class.py
## @file
#  @brief
#  Implementation of the
#  @ref DesignPatternExamples_python.handlerchain.handlerchain_message_class.Message "Message"
#  and
#  @ref DesignPatternExamples_python.handlerchain.handlerchain_message_class.MessagePosition "MessagePosition"
#  classes used in the @ref handlerchain_pattern "HandlerChain pattern"

from enum import Enum

## Type of message handled by MessageWindow.
class MessageType(Enum):
    ## Window is asked to close itself, generally sent by the window itself
    #  in response to a button up in a Close region.  Applies only to the
    #  currently selected window.
    Close = 0,

    ## Selects a window based on position.
    ButtonDown = 1,

    ## Take an action on the currently selected window.
    ButtonUp = 2


#========================================================================
#========================================================================
#========================================================================


## Position of the message in global coordinates (same scope of
#  coordinates as windows).  Although this can handle negative coordinates,
#  the example is careful to use only positive coordinates.
class MessagePosition:

    ## Constructor.
    #
    #  @param x
    #         X position to use.
    #  @param y
    #         Y position to use.
    def __init__(self, x : int, y : int) -> None:
        self.X = x
        self.Y = y

    ## @var X
    #       X position.
    #  @var Y
    #       Y position.

    ## Convert this position to a string.
    #
    #  @returns
    #     A string containing a string representation of the position.
    def ToString(self) -> str:
        return "x={0:2},y={1:2}".format(self.X, self.Y)



#========================================================================
#========================================================================
#========================================================================


## Represents a message sent to the windows.  A message contains a type
#  and a position.
class Message:

    ## Constructor.
    #
    #  @param type
    #         Value from the MessageType enumeration indicating the type of
    #         message being created.
    #  @param position
    #         Position of the message, expressed as a MessagePosition object.
    def __init__(self, type : MessageType, position : MessagePosition) -> None:
        self.MessageType = type
        self.Position = position
    
    ## @var MessageType
    ##      Value from the MessageType enumeration indicating the type of this
    #       message.
    # @var Position
    #      Position of message when the message was sent.  In a real system,
    #      this would generally represent the position of a cursor at the time
    #      the message was generated.


    ## Convert this message to a string.
    #
    #  @returns
    #     Returns a string representation of this message.
    def ToString(self) -> str:
        messageTypeAsString = ""
        match (self.MessageType):
            case MessageType.Close:
                messageTypeAsString = "Close"

            case MessageType.ButtonDown:
                messageTypeAsString = "ButtonDown"

            case MessageType.ButtonUp:
                messageTypeAsString = "ButtonUp"

            case _:
                messageTypeAsString = "Unknown message type"

        return "{0} at ({1})".format(messageTypeAsString, self.Position.ToString())

